xlml/utils/xpk.py

```diff
@@ -89,46 +89,73 @@ def _run_workload(
+    project_id: str, region: str, cluster_name: str
+) -> k8s_client.CoreV1Api:
+  """Create a core API client for the given cluster."""
   client = gke.get_authenticated_client(project_id, region, cluster_name)
 
   # Initilize the client
   core_api = k8s_client.CoreV1Api(client)
   logging.info("Successful initilize k8s client from cluster response.")
+  return core_api
+
 
-  # Get pods for the workload
+def _list_workload_pods(
+    core_api: k8s_client.CoreV1Api, workload_id: str
+) -> k8s_client.V1PodList:
+  """List all pods for the given workload."""
   logging.info(f"Getting pods for workload_id: {workload_id}")
   pods = core_api.list_namespaced_pod(
       label_selector=f"jobset.sigs.k8s.io/jobset-name={workload_id}",
       namespace="default",
   )
+  return pods
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_workload_start(
+    workload_id: str, project_id: str, region: str, cluster_name: str
+) -> bool:
+  """Check if the workload has started."""
+  core_api = _get_core_api_client(project_id, region, cluster_name)
+  pods = _list_workload_pods(core_api, workload_id)
+  print(f"Found {len(pods.items)} pods for workload {workload_id}")
+  return len(pods.items) > 0
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_workload_completion(
+    workload_id: str, project_id: str, region: str, cluster_name: str
+) -> bool:
+  """Check the workload status."""
+  core_api = _get_core_api_client(project_id, region, cluster_name)
+  pods = _list_workload_pods(core_api, workload_id)
 
-  # Check status of pods
   if not pods.items:
-    # This could happen when workload is in the queue (not initialized yet)
-    logging.info(f"No pod is found for workload selector: {workload_id}.")
+    logging.info(f"No pods found for workload selector: {workload_id}.")
+    return False
+
+  if any(pod.status.phase in ["Pending", "Running"] for pod in pods.items):
+    logging.info(f"At least one pod has yet to complete")
     return False
 
-  logging.info(f"pods: {pods}")
-  for pod in pods.items:
-    if pod.status.phase in ["Pending", "Running"]:
-      logging.info(f"One pod phase is: {pod.status.phase}")
-      return False
-    elif pod.status.phase == "Failed":
-      # Don't keep retrying if the pod has failed
-      raise AirflowFailException(f"Bad pod phase: {pod.status.phase}")
-    elif pod.status.phase in ["Unknown"]:
-      raise RuntimeError(f"Bad pod phase: {pod.status.phase}")
+  try:
+    for pod in pods.items:
+      if pod.status.phase == "Failed":
+        # Don't keep retrying if the pod has failed
+        raise AirflowFailException(f"Bad pod phase: {pod.status.phase}")
+      elif pod.status.phase in ["Unknown"]:
+        raise RuntimeError(f"Bad pod phase: {pod.status.phase}")
+  finally:
+    # Print the logs of the last pod checked - either the first failed pod or
+    # the last successful one.
+    logs = core_api.read_namespaced_pod_log(
```

Comment:
Do you have an example I could take a quick look?

Reply:
I'll deploy to the dev env with a successful run and one with a bad workload to show the failure case.

Reply:
Here is a link to a successful run: https://fc137785b6c943a4af82d0671bf0ebe5-dot-us-central1.composer.googleusercontent.com/dags/gke_example_dag/grid?root=&dag_run_id=manual__2024-03-26T19%3A31%3A59.601549%2B00%3A00&task_id=resnet-single-slice-v4-8.run_model.wait_for_workload_completion&tab=logs

I'll follow up with the failure case.

Reply:
Thanks! Do you know why this task is not scheduled to run? https://screenshot.googleplex.com/9UPHdZx8oAqwbkz

Reply:
Ah it's a bit outdated, I hadn't added the GCS support here. I'll do another run shortly.